### ISSUE TITLE
Extraction from optional root

### ISSUE THREAD

**Issue by fonkadelic** [pointfreeco/swift-case-paths#40]
Hi,
updating from `0.2.0` to `0.5.0` resulted in an unexpected behavior for me. It seems that is no longer possible to extract a CasePath from from an optional value. 
The following test case works fine in `0.2.0`, but fails on `0.5.0`:

```
  func testExtractFromOptionalRoot() {
    enum Authentication {
      case authenticated(token: String)
      case unauthenticated
    }
    
    // Fails
    let optionalAuthentication: Authentication? = .authenticated(token: "deadbeef")
    XCTAssertNotNil(extract(case: Authentication.authenticated, from: optionalAuthentication))
  }
```

It has probably something to do with moving away from `Mirror`, but i didn't dig deeper.

Would be good to know if you are aware of this change in the behavior and if it's intended.

**Comment by stephencelis** [comment]
@fonkadelic Dang, this is one of those cases where optional promotion isn't doing us any favors. To work around the issue we'd recommend unwrapping the optional first:

```swift
XCTAssertNotNil(
  optionalAuthentication.flatMap { extract(case: Authentication.authenticated, from: $0) }
)
```

Or if you're using a case path, you can prepend it with `/Optional.some`.

Given Swift's behavior, we should probably support this again. We probably won't dive into a fix immediately, but will try to look into it soon. @mayoff has also been in the code recently and might know a quick fix.

**Comment by mayoff** [comment]
Here's the signature of `extract`:

```
public func extract<Root, Value>(case embed: @escaping (Value) -> Root, from root: Root) -> Value?
```

The test case passes a `root` defined in a prior statement, so its type is fixed. Swift therefore adds a constraint `Root == Optional<Authentication>` when type-checking the statement that calls `extract`.

The unconstrained type of `Authentication.authenticated` is `(String) -> Authentication`. Swift converts this to automatically to `(String) -> Optional<Authentication>` to satisfy the constraint.

This would be pretty hard to fix. It's not too hard to detect though, so I'll assert if it's detected.

**Comment by lukeredpath** [comment]
I was just coming here to post the same bug after trying to update to CasePaths 0.5.0 (from 0.4.0 in my case), where our feature state has an optional `enum` value and I'm trying to extract the enum value in a pullback. It would be great if this could be fixed, but good to see there's a workaround.

**Comment by mayoff** [comment]
I just pushed another commit that should return the correct value instead of asserting in this scenario.

**Comment by lukeredpath** [comment]
@mayoff Sterling work, thank you. I will update our integration branch and try this tomorrow.

**Comment by lukeredpath** [comment]
@mayoff I just tried this out in our app and it seems to have fixed the issue with our pullbacks, so that's great.

I found one other issue that I was able to work around, but is a change in behaviour from 0.4.0. We have an interface that returns a `Result<T?, Error>` which we were using in a test assertion which is now failing unless I unwrap the optional. For example, given something like this:

```swift
func fetch() -> Result<String?, Error> {
  return .success("hello world")
}
```

The following assertion now fails, but used to pass:

```swift
XCTAssertEqual(
    "hello world",
    extract(case: Result.success, from: fetch())
)
```

However, this does pass:

```swift
XCTAssertEqual(
    .some("hello world"),
    extract(case: Result.success, from: fetch())
)
```

**Comment by mayoff** [comment]
@lukeredpath I think you should go ahead and open another issue for that problem. [This commit](https://github.com/mayoff/swift-case-paths/commit/9154dda26adda61ab2612db9e9a5a4babe924c7c) should fix it but this issue and pull request #41 are already closed.
